### PR TITLE
Glicko2

### PR DESCRIPTION
**Pull request by paiqu**
- Combine changes from FR2-1 and FR2-2
- Add structure to run scheduled tasks that will be used to run the Glicko-2 update function

### CODE REVIEW

File: src/elo/utils/matchBracket.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    } else if (rating >= eloConstants.PLATINUMBASE && rating < eloConstants.DIAMONDBASE) {
+        return 'Platinum';
+    } else if (rating >= eloConstants.DIAMONDBASE && rating < eloConstants.CHAMPIONBASE) {
+        return 'Diamond';
+    } else {
+        return 'Champion';
+    }
+}
+
+export default matchBracket;
```
[the remark this answers is not in the document]
Comment (vck3000) [reply src/elo/utils/matchBracket.js:23]: Might be the other way around depending if the base is the min value of that category. Thanks @FakhruddinNalawala !

File: src/gameplay/glicko2.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+              g.winningTeam === TeamEnum.RESISTANCE
+                ? OutcomeEnum.WIN
+                : OutcomeEnum.LOSE,
+          };
+        }
+      }),
+    );
+  }
+
+  static async computeRankRatingsByUserId(userId: string): Promise<IRank> {
```
Comment (vck3000) [inline src/gameplay/glicko2.ts:129]: Move this up!

File: src/gameplay/game.ts [hunk 3]
```diff
@@ -1366,8 +1369,10 @@ class Game extends Room {
       const ratingRecordToStore = {
         timeGameFinished: timeFinished,
         winningTeam: this.winner,
-        spyTeam: this.spyUsernames,
-        resistanceTeam: this.resistanceUsernames,
+        spyTeam: this.spyUsernames.map((str) => str.toLowerCase()),
+        resistanceTeam: this.resistanceUsernames.map((str) =>
+          str.toLowerCase(),
+        ),
```
Comment (vck3000) [inline src/gameplay/game.ts:1375]: @FakhruddinNalawala Please give this a test and let me know if it's good!

File: src/elo/constants/eloConstants.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const eloConstants = {
+    // ELO System Constants
+    DEFAULTRATING: 1500,
```
[the remark this answers is not in the document]
Comment (paiqu) [reply src/elo/constants/eloConstants.ts]: @cwei9203

File: src/elo/constants/eloConstants.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    DEFAULTRD: 350,
+    DEFAULTVOL: 0.06,
+    PROVISIONGAMES : 20,
+    BRONZEBASE : 1300,
+    SILVERBASE : 1400,
+    GOLDBASE : 1550,
+    PLATINUMBASE : 1700,
+    DIAMONDBASE : 1800,
+    CHAMPIONBASE : 1900,
```
[the remark this answers is not in the document]
Comment (paiqu) [reply src/elo/constants/eloConstants.ts]: @cwei9203

File: src/elo/utils/getSeasonNumber.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+        //     }
+        // });
+        const returnedSeasonNumber = await seasonNumber.findOne({}).exec();
+        return returnedSeasonNumber.number;
+    } catch (err) {
+        console.error(err);
+    }
+}
+
+export default getSeasonNumber;
```
[the remark this answers is not in the document]
Comment (paiqu) [reply src/elo/utils/getSeasonNumber.js:17]: @cwei9203

File: src/elo/utils/getSeasonNumber.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import seasonNumber from '../../models/seasonNumber.js';
+
+async function getSeasonNumber() {
+    try {
+        // seasonNumber.findOne({}).exec().then((returnedSeasonNumber) => {
+        //     if (!returnedSeasonNumber) {
+        //         seasonNumber.create({ number: 1 });
+        //     }
+        // });
```
[the remark this answers is not in the document]
Comment (paiqu) [reply src/elo/utils/getSeasonNumber.js:9]: @cwei9203

File: src/elo/utils/matchBracket.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (paiqu) [reply src/elo/utils/matchBracket.js:23]: @cwei9203

File: src/app.ts [hunk 8]
```diff
@@ -177,8 +178,9 @@ app.use('/profile', profileRoutes);
 const IP = process.env.IP || '127.0.0.1';
 const server = app.listen(port, () => {
   console.log(`Server has started on ${IP}:${port}!`);
-  const glickoUpdateSchedule = schedule.scheduleJob('0 0 * * *', () => {
+  const glickoUpdateSchedule = schedule.scheduleJob('0 0 * * *', async () => {
     console.log('Running scheduled work ');
+    // await Glicko2.updateAllUsersRatings();
   });
```
Comment (vck3000) [inline src/app.ts:184]: Let's pull this out. Maybe put it at the bottom of this file.

File: src/db/mongo.ts [hunk 9]
```diff
@@ -6,6 +6,10 @@ import type { IRatingPeriodGameRecord } from '../models/types';
 import { IRank, IUser } from '../models/types';
 
 export default class Mongo {
+  static getAllUsers(): Promise<IUser[]> {
+    return User.find({}).exec();
+  }
```
Comment (vck3000) [inline src/db/mongo.ts:11]: Yep, let's move this to use eachAsync somehow :)

File: src/db/mongo.ts [hunk 10]
```diff
@@ -25,20 +29,53 @@ export default class Mongo {
     return user;
   }
 
-  static getGamesByUsername(
+  static async getGamesByUsername(
```
Comment (vck3000) [inline src/db/mongo.ts:32]: Rm async if it's ok :)

File: src/db/mongo.ts [hunk 11]
```diff
@@ -25,20 +29,53 @@ export default class Mongo {
...
-    username = username.toLowerCase();
+    const usernameLower = username.toLowerCase();
 
     return RatingPeriodGameRecord.find({
-      $or: [{ spyTeam: username }, { resistanceTeam: username }],
+      $or: [{ spyTeam: usernameLower }, { resistanceTeam: usernameLower }],
     }).exec();
   }
 
   static async getRankByUserId(userId: string): Promise<IRank> {
```
Comment (vck3000) [inline src/db/mongo.ts:42]: Maybe let's rename this to `getUserRankById`.

File: src/db/mongo.ts [hunk 12]
```diff
@@ -25,20 +29,53 @@ export default class Mongo {
...
+    }).exec();
+
+    if (!rank) {
+      throw Error(`Could not find rank for user id: ${userId}`);
+    }
+
+    return rank;
+  }
+
+  static async getRankByUsername(username: string): Promise<IRank> {
```
Comment (vck3000) [inline src/db/mongo.ts:55]: getUserRankByUsername

File: src/elo/constants/eloConstants.ts [hunk 13]
```diff
@@ -1,15 +1,15 @@
...
-    BRONZEBASE : 1300,
-    SILVERBASE : 1400,
-    GOLDBASE : 1550,
-    PLATINUMBASE : 1700,
-    DIAMONDBASE : 1800,
-    CHAMPIONBASE : 1900,
+  // ELO System Constants
+  DEFAULT_RATING: 1500,
+  DEFAULT_RD: 350,
+  DEFAULT_VOL: 0.06,
```
Comment (vck3000) [inline src/elo/constants/eloConstants.ts:5]: Who was pulling this out? The glicko-2 constants vs the other elo bracket constants.

File: src/elo/utils/matchBracket.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+        return 'Silver';
+    } else if (rating < eloConstants.PLATINUM_BASE) {
+        return 'Gold';
+    } else if (rating < eloConstants.DIAMOND_BASE) {
+        return 'Platinum';
+    } else if (rating < eloConstants.CHAMPION_BASE) {
+        return 'Diamond';
+    } else {
+        return 'Champion';
+    }
```
Comment (vck3000) [inline src/elo/utils/matchBracket.ts:20]: Let's reverse the order here.

File: src/elo/utils/matchBracket.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import eloConstants from "../constants/eloConstants";
+
+function matchBracket(rating:number, totalRankedGamesPlayed:number) {
```
Comment (vck3000) [inline src/elo/utils/matchBracket.ts:3]: Should probably call this something better.

File: src/gameplay/glicko2.ts [hunk 16]
```diff
@@ -6,6 +6,103 @@ class Glicko2 {
...
+      username: user.username,
+      mu: (userRankData.playerRating - 1500) / 173.7178,
+      phi: userRankData.rd / 173.7178,
+      ratingVolatility: userRankData.volatility,
+    };
+
+    // Check if the player competed during the rating period
+    // If the player does not compete during the rating period
+    // Rating and Volatility remain the same, but the RD increases
+    if (gameSummary.length == 0) {
```
Comment (vck3000) [inline src/gameplay/glicko2.ts:27]: `if (!gameSummary.length) {`

File: src/modelsHelper/seasonNumber.ts [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+export async function incrementSeasonNumber(): Promise<number> {
+  const result = await SeasonNumber.findOneAndUpdate(
+    {},
+    { $inc: { number: 1 } },
+    { new: true },
+  );
+
+  return result.number;
+}
```
Comment (vck3000) [inline src/modelsHelper/seasonNumber.ts:22]: This should be in the db/ folder

File: src/db/mongo.ts (hunk 9 above)
Comment (vck3000) [reply src/db/mongo.ts:11]: Let's remove this.

File: src/db/mongo.ts [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+      .find({})
+      .cursor()
+      .eachAsync(async (user: IUser) => {
+        const updatedRank = await updateFn(user._id.toString());
+        await this.updateRankRatings(user._id.toString(), updatedRank);
+      });
+  }
+
+  static getUserByUsername(username: string): Promise<IUser> {
+    const user = User.findOne({ usernameLower: username.toLowerCase() }).exec();
```
Comment (vck3000) [inline src/db/mongo.ts:24]: await?

File: src/db/mongo.ts [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    if (!user) {
+      throw Error(`Could not find user: ${username}`);
+    }
+
+    return user;
+  }
+
+  static getUserByUserId(userId: string): Promise<IUser> {
+    const user = User.findOne({ _id: new Types.ObjectId(userId) }).exec();
```
Comment (vck3000) [inline src/db/mongo.ts:33]: await

File: src/db/mongo.ts [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+    if (!user) {
+      throw Error(`Could not find user id: ${userId}`);
+    }
+
+    return user;
+  }
+
+  static getGamesByUsername(
```
Comment (vck3000) [inline src/db/mongo.ts:42]: getRatingPeriodGames?

File: src/db/mongo.ts [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    const usernameLower = username.toLowerCase();
+
+    return RatingPeriodGameRecord.find({
+      $or: [{ spyTeam: usernameLower }, { resistanceTeam: usernameLower }],
+    }).exec();
+  }
+
+  static async getUserRankByUserId(userId: string): Promise<IRank> {
+    const user = await this.getUserByUserId(userId);
+    const rank = Rank.findOne({
```
Comment (vck3000) [inline src/db/mongo.ts:54]: await

File: src/db/mongo.ts [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    return rank;
+  }
+
+  static async getUserRankByUsername(username: string): Promise<IRank> {
+    const usernameLower = username.toLowerCase();
+
+    const user = await this.getUserByUsername(usernameLower);
+    const rank = Rank.findOne({
```
Comment (vck3000) [inline src/db/mongo.ts:69]: await

File: src/gameplay/types.ts [hunk 23]
```diff
@@ -27,39 +26,12 @@ export interface See {
...
-  mutedPlayers?: string[];
-  patreonId?: string;
-  IPAddresses?: string[];
-  lastIPAddress?: string;
+export enum TeamEnum {
+  RESISTANCE = 'Resistance',
+  SPY = 'Spy',
+}
+
+export enum OutcomeEnum {
```
Comment (vck3000) [inline src/gameplay/types.ts:34]: Let's call these Team and Outcome

**Review by vck3000 — COMMENTED** [review]
Looks good!

Give it a final manual test and we can merge!

**Comment by vck3000** [comment]
LGTM! I'll give it a test sometime tonight/tmr night :)

**Pull request closed without merging**